atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/async/client/creation/DefaultCqlClientFactory.java

```diff
@@ -0,0 +1,128 @@
+
+public final class DefaultCqlClientFactory implements CqlClientFactory {
+    public static final CqlClientFactory DEFAULT = new DefaultCqlClientFactory();
+
+    private static final Logger log = LoggerFactory.getLogger(DefaultCqlClientFactory.class);
+    private static final String LOAD_BALANCING_POLICY = "DcInferringLoadBalancingPolicy";
+    private static final String COMPRESSION_PROTOCOL = "lz4";
+
+    private final Supplier<CqlSessionBuilder> cqlSessionBuilderFactory;
+
+    public DefaultCqlClientFactory(Supplier<CqlSessionBuilder> cqlSessionBuilderFactory) {
+        this.cqlSessionBuilderFactory = cqlSessionBuilderFactory;
+    }
+
+    public DefaultCqlClientFactory() {
+        this(CqlSessionBuilder::new);
+    }
+
+    @Override
+    public CqlClient constructClient(
+            TaggedMetricRegistry taggedMetricRegistry,
+            CassandraKeyValueServiceConfig config,
+            boolean initializeAsync) {
+        return config.servers().accept(new CassandraServersConfigs.Visitor<CqlClient>() {
+            @Override
+            public CqlClient visit(CassandraServersConfigs.DefaultConfig defaultConfig) {
+                return ThrowingCqlClientImpl.INSTANCE;
+            }
+
+            @Override
+            public CqlClient visit(CqlCapableConfig cqlCapableConfig) {
+                if (!cqlCapableConfig.thriftAndCqlHostsMatch()) {
+                    log.warn("Your CQL capable config is wrong, the hosts for CQL and Thrift are not the same, using "
+                            + "async API will result in an exception.");
+                    return ThrowingCqlClientImpl.INSTANCE;
+                }
+
+                Set<InetSocketAddress> servers = cqlCapableConfig.cqlHosts();
+                DriverConfigLoader driverConfigLoader = buildDriverConfigLoader(config);
+                Supplier<CqlSession> cqlSessionSupplier =
+                        constructCqlSessionFactory(servers, driverConfigLoader, config);
+
+                return CqlClientImpl.create(
+                        taggedMetricRegistry,
+                        cqlSessionSupplier,
+                        cqlCapableConfig.tuning(),
+                        initializeAsync);
+            }
+        });
+    }
+
+    private DriverConfigLoader buildDriverConfigLoader(CassandraKeyValueServiceConfig config) {
+        return DriverConfigLoader.programmaticBuilder()
+                .withString(DefaultDriverOption.LOAD_BALANCING_POLICY_CLASS, LOAD_BALANCING_POLICY)
+                .withString(DefaultDriverOption.PROTOCOL_COMPRESSION, COMPRESSION_PROTOCOL)
+                .withInt(DefaultDriverOption.CONNECTION_POOL_LOCAL_SIZE, config.poolSize())
+                .withInt(DefaultDriverOption.CONNECTION_POOL_REMOTE_SIZE, config.poolSize())
+                .withInt(DefaultDriverOption.REQUEST_PAGE_SIZE, config.fetchBatchCount())
+                .build();
+    }
+
+    private Supplier<CqlSession> constructCqlSessionFactory(
```

Comment:
nit/non-actionable: It may be easier to just have this return `CqlSession` and wrap it in a lambda in the function, though I don't mind either way.

Reply:
I have to rebase some of these changes. Will take another look after that.

Reply:
Not an issue in the 3.x version.